 <a name="TFM"></a> 

# Unsupervised Clustering

This repository contains a Shiny [user interface](https://github.com/aalvarogaarcia/TFM/blob/main/UI/user-interface.md) and an [R Markdown](https://github.com/aalvarogaarcia/TFM/tree/main/RMD/RMD.md) report for unsupervised clustering for endophenotype discovery in genetic and proteomics data. 
The project was developed as part of a Master's thesis in Bioinformatics and biostatistics from UOC-UB in collaboration with the Genomics Division of the Institute of Technology and Renewable Energy (https://www.iter.es/areas/area-genomica/).

## Table of contents

* [Objective](#objective)
* [Methodology](#methodology)
* [Target Audience](#target-audience)
* [Disclaimer](#disclaimer)
* [Next Steps](#next-steps)

## Objective

The main objective of this project was to develop a user-friendly tool for exploring and analyzing genetic and protein data using unsupervised clustering techniques. 
The tool should allow users to:

* Perform univariate and bivariate analysis of genetic and protein data.
* Conduct multivariate analysis to identify endotypes using various clustering algorithms.
* Visualize the results of the clustering analysis using different visualization techniques.


## Methodology

The project was developed using R and Shiny. The Shiny user interface provides a user-friendly way to interact with the data and perform the analysis. 
The R Markdown report provides a detailed overview of the methodology and results.

### Key Features:

* Interactive Shiny user interface: The Shiny user interface allows users to easily upload data, select clustering algorithms and visualization techniques, and view the results of the analysis.
* Comprehensive R Markdown report: The R Markdown report provides a detailed overview of the methodology, including data preprocessing, clustering analysis, and visualization.
* Multiple R Scripts: In these scripts a whole series of methods and functions are developed with different objectives for the analysis and visualization of data.

## Target Audience:

 This tool is intended for researchers and students in the fields of bioinformatics, genetics, and genomics who are interested in analyzing genetic and protein data using unsupervised clustering techniques.

## Disclaimer:

 This tool is for research purposes only and should not be used for clinical diagnosis or treatment.

## Next Steps:

 Future work could include:

* Extending the tool to support additional data types.
* Developing more advanced visualization techniques.
* Integrating the tool with other bioinformatics tools.


<hr>
<!-- ------------------ SECTION 7 ------------------ -->

<a name="References"></a>
## References

<ol>
<li><a href="https://www.who.int/publications/i/item/9789240018440">Genomic sequencing of SARS-CoV-2. A guide to implementation for maximum impact on public health</a>, WHO, January 8, 2021.</li>
<li><a href="https://apps.who.int/iris/handle/10665/3"> Report “Global genomic surveillance strategy for pathogens with pandemic and epidemic potential, 2022-2032”</a>. Ginebra, WHO, 2022.</li>
<li>Gohl DM, Garbe J, Grady P, et al. A rapid, cost-effective tailed amplicon method for sequencing SARS-CoV-2. <i>BMC Genomics</i>. 2020;21(1):863. Published 2020 Dec 4. <a href="https://doi.org/10.1186/s12864-020-07283-6">doi:10.1186/s12864-020-07283-6</a>.</li>
<li>Itokawa K, Sekizuka T, Hashino M, Tanaka R, Kuroda M. Disentangling primer interactions improves SARS-CoV-2 genome sequencing by multiplex tiling PCR. <i>PLoS One</i>. 2020;15(9):e0239403. Published 2020 Sep 18. <a href="https://doi.org/10.1371/journal.pone.0239403">doi:10.1371/journal.pone.0239403</a>.</li> 
<li>Koskela von Sydow A, Lindqvist CM, Asghar N, et al. Comparison of SARS-CoV-2 whole genome sequencing using tiled amplicon enrichment and bait hybridization. <i>Sci Rep</i>. 2023;13(1):6461. Published 2023 Apr 20. <a href="https://doi.org/10.1038%2Fs41598-023-33168-1">doi:10.1038/s41598-023-33168-1</a>.</li>
<li>Zhou B, Wentworth DE. Influenza A virus molecular virology techniques. <i>Methods Mol Biol.</i> 2012;865:175-192. doi:<a href="https://link.springer.com/protocol/10.1007/978-1-61779-621-0_11">10.1007/978-1-61779-621-0_11</a>.</li>
<li>Zhou B, Lin X, Wang W, et al. Universal influenza B virus genomic amplification facilitates sequencing, diagnostics, and reverse genetics. <i>J Clin Microbiol</i>. 2014;52(5):1330-1337. doi:<a href="https://doi.org/10.1128/jcm.03265-13">10.1128/JCM.03265-13</a>.</li>
<li>Ying Lin, Jeffrey Koble, Priyanka Prashar, Anita Pottekat, Christina Middle, Scott Kuersten, Michael Oberholzer, Robert Brazas, Darcy Whitlock, Robert Schlaberg, Gary P. Schroth. A sequencing and subtyping protocol for Influenza A and B viruses using Illumina® COVIDSeq™ Assay Kit. Protocols.io. doi:<a href="https://dx.doi.org/10.17504/protocols.io.n2bvj8mrxgk5/v1">dx.doi.org/10.17504/protocols.io.n2bvj8mrxgk5/v1
</a></li>
</ol>
  
<p align="right">
  <a href="#Influenza" title="Up">
    <img src="https://github.com/genomicsITER/influenza/blob/main/images/home-icon.png" style="float: right; margin: 10px; padding: 2px;" />
  </a>
</p>



<hr>
<!-- ------------------ SECTION 8 ------------------ -->

<a name="Acknowledgements"></a>
## Acknowledgements

The development of this Master's Thesis and definitely this repository has been possible thanks to the invaluable support of the team of the research group of the Genomics Area of the Instituto Tecnológico y de Energías Renovables (ITER), in particular my external tutor José Miguel Lorenzo Salazar, for his invaluable guidance. This work would also not have been possible without the help of my UOC tutor, Dr. José Luis Mosquera Mayo, for his follow-up and valuable suggestions. To all of them, my sincere thanks.

<p align="right">
  <a href="#Influenza" title="Up">
    <img src="https://github.com/genomicsITER/influenza/blob/main/images/home-icon.png" style="float: right; margin: 10px; padding: 2px;" />
  </a>
</p>



<hr>
<!-- ------------------ SECTION 11 ------------------ -->

<a name="How-to-cite"></a>
## How to cite this work

> This work has not been publised yet. See 'License and Attribution' section to cite this repository.

> To use the deposited sequences at GISAID, please, acknowledge this work as recommended by GISAID. Find the 'GISAID acknowledge tables' <a href="https://github.com/genomicsITER/influenza/tree/main/sequences/acknowledgements">here</a>.

<p align="right">
  <a href="#Influenza" title="Up">
    <img src="https://github.com/genomicsITER/influenza/blob/main/images/home-icon.png" style="float: right; margin: 10px; padding: 2px;" />
  </a>
</p>


<hr>
<!-- ------------------ SECTION 9 ------------------ -->

<a name="License and Attribution"></a>
## License and Attribution ![Copyright](https://github.com/aalvarogaarcia/TFM/blob/main/images/ccbyncnd.png)

Attribution-NonCommercial-NoDerivatives (CC BY-NC-ND) License
This project is licensed under the Creative Commons Attribution-NonCommercial-NoDerivatives 4.0 International License.

You are free to Share (copy and redistribute the material in any medium or format) with Attribution (give appropriate credit, provide a link to the license, and indicate if changes were made).
You may not use this for commercial purposes.
You may not Adapt (remix, transform, or build upon the material).
For the full license text, please refer to the Creative Commons Attribution-NonCommercial-NoDerivatives 4.0 International License.

If you want to use this project commercially or create derivative works, please contact me for permission.
<p align="right">
  <a href="#Influenza" title="Up">
    <img src="https://github.com/genomicsITER/influenza/blob/main/images/home-icon.png" style="float: right; margin: 10px; padding: 2px;" />
  </a>
</p>




<hr>
<!-- ------------------ SECTION 12 ------------------ -->

<a name="Update logs"></a>
## Update logs

> June 18, 2024. My TFM github repository became public!

> May 19, 2024. Working hard into my GH repository.

> May 3, 2024. The developed UI is now available here.
 
> March 1, 2024. Created the private version of this repository.

<p align="right">
  <a href="#TFM" title="Up">
    <img src="https://github.com/aalvarogaarcia/TFM/blob/main/images/home-icon.png" style="float: right; margin: 10px; padding: 2px;" />
  </a>
</p>
